xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -444,6 +555,177 @@ public void run() {
+      Status abortStatus = null;
+      Integer abortRate = null;
+      if (faultAbort.headerAbort()) {
+        try {
+          String httpCodeStr = headers.get(HEADER_ABORT_HTTP_STATUS_KEY);
+          if (httpCodeStr != null) {
+            int httpCode = Integer.parseInt(httpCodeStr);
+            abortStatus = GrpcUtil.httpStatusToGrpcStatus(httpCode);
+          }
+          String grpcCodeStr = headers.get(HEADER_ABORT_GRPC_STATUS_KEY);
+          if (grpcCodeStr != null) {
+            int grpcCode = Integer.parseInt(grpcCodeStr);
+            abortStatus = Status.fromCodeValue(grpcCode);
+          }
+          String abortPercentageStr = headers.get(HEADER_ABORT_PERCENTAGE_KEY);
+          if (abortPercentageStr != null) {
+            int abortPercentage =
+                Integer.parseInt(headers.get(HEADER_ABORT_PERCENTAGE_KEY));
+            abortRate = abortPercentage * 100 * 100;
+          }
+        } catch (NumberFormatException e) {
+          return null; // treated as header_abort not applicable
+        }
+      } else {
+        abortStatus = faultAbort.status();
+      }
+      if (abortRate == null) {
+        abortRate = faultAbort.ratePerMillion();
+      }
+      if (random.nextInt(1_000_000) >= abortRate) {
+        return null;
+      }
+      return abortStatus;
+    }
+  }
+
+  /**
+   * A forwarding client call that counts active fault injections.
+   */
+  private final class ActiveFaultCountingClientCall<ReqT, RespT> extends
+      SimpleForwardingClientCall<ReqT, RespT> {
+    ActiveFaultCountingClientCall(ClientCall<ReqT, RespT> faultInjectedDelegate) {
+      super(faultInjectedDelegate);
+      activeFaultInjectedStreams.incrementAndGet();
+    }
+
+    @Override
+    public void start(Listener<RespT> listener, Metadata headers) {
+      listener = new SimpleForwardingClientCallListener<RespT>(listener) {
+        @Override
+        public void onClose(Status status, Metadata trailers) {
+          delegate().onClose(status, trailers);
+          activeFaultInjectedStreams.decrementAndGet();
+        }
+      };
+      delegate().start(listener, headers);
+    }
+  }
+
+  /** A {@link DelayedClientCall} with a fixed delay. */
+  private static final class DelayInjectedCall<ReqT, RespT> extends DelayedClientCall<ReqT, RespT> {
+    final Object lock = new Object();
+    ScheduledFuture<?> delayTask;
+    boolean cancelled;
+
+    DelayInjectedCall(
+        long delayNanos, Executor callExecutor, ScheduledExecutorService scheduler,
+        @Nullable Deadline deadline,
+        final Supplier<? extends ClientCall<ReqT, RespT>> callSupplier) {
+      super(callExecutor, scheduler, deadline);
+      ScheduledFuture<?> task = scheduler.schedule(
+          new Runnable() {
+            @Override
+            public void run() {
+              setCall(callSupplier.get());
+            }
+          },
+          delayNanos,
+          NANOSECONDS);
```

Comment:
s/NANOSECONDS/TimeUnit.NANOSECONDS/g

Reply:
go/java-practices/imports
Otherwise when you import to google, the linter will ask you to change back.

Reply:
Really?  That says "The method/field should be fairly well-known, and self-describing". Most of existing usages (if not all?) in our codebase do not statically import the enum value.

Reply:
I think NANOSECONDS is fairly well-known, and self-describing. It's a time unit for sure.

This is one of suggestions found in import:
```
JavaOptionalSuggestions
StaticImportPreferred
Aug 6, 2020, 10:59:10 AM
A static import is usually clearer for these common APIs. Note that in this case, statically importing two methods with the same name is perfectly safe: the compiler will simply perform overload resolution on the methods and select the most appropriate one.

Suggested replacement:
      remainingNanos = min(remainingNanos, deadline.timeRemaining(NANOSECONDS));
    }
    if (contextDeadline != null && contextDeadline.timeRemaining(NANOSECONDS) < remainingNanos) {
      remainingNanos = contextDeadline.timeRemaining(NANOSECONDS);
      if (logger.isLoggable(Level.FINE)) {
        StringBuilder builder =
            new StringBuilder(
                String.format(
                    "Call timeout set to '%d' ns, due to context deadline.", remainingNanos));
        if (deadline == null) {
... (click "show fix" for full suggestion)

http://go/java-practices/imports
```

Reply:
Actually we don't care too much about this, as different IDE setting tend to mess it up.